### PR TITLE
Allow user defined dbt target_path config for dbt run and test solids

### PR DESCRIPTION
**Pull request by makotonium**
## Summary
Since dbt allows customization of `target_path` in your project file (`dbt_project.yml`), Dagster integration should allow customization of the path upon configuring dbt solids. This PR introduces a new configuration called `target_path` for `dbt_cli_test` and `dbt_cli_run`. 

The motivation behind this change was from being able to run dbt model and test on a clone of target database prior to deploying to the target database along with using [their state method](https://docs.getdbt.com/reference/node-selection/methods#the-state-method) to only run models that have been modified. (Need to keep the target directories separate for a test run and actual deployment to keep track of diffs)

[Original Slack thread](https://dagster.slack.com/archives/C01U954MEER/p1621363140015400)

## Test Plan
- Updated `libraries/dagster-dbt/dagster_dbt_tests/cli/test_slids.py` for `dbt run` and `dbt test`

## Checklist

- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by makotonium** [comment]
@OwenKephart I tried to run tests locally for `.../dagster-dbt/dagster_dbt_tests/cli/test_solids.py`, but they are getting skipped with the following errors (even prior to my changes). Any tips on getting them to run or another way of verifying changes? 

```
dagster_dbt_tests/cli/test_solids.py::TestDbtCliSolids::test_dbt_cli_with_unset_env_var_in_profile
  /Users/makoto/src/dagster/python_modules/libraries/dagster-dbt/dagster_dbt_tests/conftest.py:45: RuntimeWarning: Error launching Dockerized Postgres: Failed to launch docker container(s) via docker-compose: Traceback (most recent call last):
    File "urllib3/connectionpool.py", line 670, in urlopen
    File "urllib3/connectionpool.py", line 392, in _make_request
    File "http/client.py", line 1255, in request
    File "http/client.py", line 1301, in _send_request
    File "http/client.py", line 1250, in endheaders
    File "http/client.py", line 1010, in _send_output
    File "http/client.py", line 950, in send
    File "docker/transport/unixconn.py", line 43, in connect
  FileNotFoundError: [Errno 2] No such file or directory
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "requests/adapters.py", line 439, in send
    File "urllib3/connectionpool.py", line 726, in urlopen
    File "urllib3/util/retry.py", line 410, in increment
    File "urllib3/packages/six.py", line 734, in reraise
    File "urllib3/connectionpool.py", line 670, in urlopen
    File "urllib3/connectionpool.py", line 392, in _make_request
    File "http/client.py", line 1255, in request
    File "http/client.py", line 1301, in _send_request
    File "http/client.py", line 1250, in endheaders
    File "http/client.py", line 1010, in _send_output
    File "http/client.py", line 950, in send
    File "docker/transport/unixconn.py", line 43, in connect
  urllib3.exceptions.ProtocolError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "docker/api/client.py", line 214, in _retrieve_server_version
    File "docker/api/daemon.py", line 181, in version
    File "docker/utils/decorators.py", line 46, in inner
    File "docker/api/client.py", line 237, in _get
    File "requests/sessions.py", line 543, in get
    File "requests/sessions.py", line 530, in request
    File "requests/sessions.py", line 643, in send
    File "requests/adapters.py", line 498, in send
  requests.exceptions.ConnectionError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "docker-compose", line 3, in <module>
    File "compose/cli/main.py", line 81, in main
    File "compose/cli/main.py", line 199, in perform_command
    File "compose/cli/command.py", line 60, in project_from_options
    File "compose/cli/command.py", line 152, in get_project
    File "compose/cli/docker_client.py", line 41, in get_client
    File "compose/cli/docker_client.py", line 170, in docker_client
    File "docker/api/client.py", line 197, in __init__
    File "docker/api/client.py", line 221, in _retrieve_server_version
  docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
  [80710] Failed to execute script docker-compose
  
    {"hostname": postgres_host} if IS_BUILDKITE else {},

-- Docs: https://docs.pytest.org/en/stable/warnings.html

```

File: python_modules/libraries/dagster-dbt/dagster_dbt/cli/solids.py [hunk 1]
```diff
@@ -190,7 +205,8 @@ def dbt_cli_run(context) -> DbtCliOutput:
         warn_error=context.solid_config["warn-error"],
         ignore_handled_error=context.solid_config["ignore_handled_error"],
     )
-    run_results = parse_run_results(context.solid_config["project-dir"])
+    target_path = __dbt_target_path(context.solid_config["target"])
```
Comment (OwenKephart) [inline python_modules/libraries/dagster-dbt/dagster_dbt/cli/solids.py:208]: I think this should be context.solid_config["target-path"]. Also I think you can get rid of the __dbt_target_path function and just give the config field a default of "target" (`default_value="target"`, https://docs.dagster.io/_apidocs/config#dagster.Field)

**Review by OwenKephart — CHANGES_REQUESTED** [review]
Overall (beyond the minor comment), looks good! I'll look into the test failure issue and see what's going on there.

**Comment by OwenKephart** [comment]
I think once you implement the changes I mention, the errors _should_ go away. There's a linting issue with the __dbt_target_path function (the else block doesn't have a return statement on it), but that should go away once you remove the function.

Also make sure to run `make black` at the top level of the repo before pushing to get rid of any formatting inconsistencies.

File: python_modules/libraries/dagster-dbt/dagster_dbt/cli/solids.py (hunk 1 above)
Comment (makotonium) [reply python_modules/libraries/dagster-dbt/dagster_dbt/cli/solids.py:208]: That's a much cleaner implementation to use the `default_value` config. I didn't know about it. Thanks!

**Comment by makotonium** [comment]
@OwenKephart Well it turns out the errors I mentioned earlier were from me not having Docker Desktop running. 🤦 

I made some changes based on your feedback. Thank you. Test are passing locally now. But I did have to do `pip install dbt` to make tests run, which should maybe documented in the `README` for this package?

I'd like to make a few documentation changes so I'll push a few more commits soon.

**Comment by OwenKephart** [comment]
Hi @makotonium thanks so much for your contribution! Everything looks good, so I'll go ahead and merge this in for you. Feel free to send any future diffs my way (we always appreciate it!)

**Comment by makotonium** [comment]
@OwenKephart You're welcome! I just came back to this PR to see if all the CI tests passed and I saw a few failures from `dagster-io.dagster` jobs but you were too fast for me. 😄 Not sure if the errors were related to my changes...

I wasn't quite done in terms of updating the documentation but glad it looked OK. Thanks for your helpful feedback and helping me through. I'll just create another PR for documentation updates.

**Comment by OwenKephart** [comment]
Oh yep sorry for jumping the gun on you haha, assumed you were planning on creating a separate PR for the documentation stuff -- those failures were not relevant (just weren't able to launch).

**Pull request merged**